hazelcast/src/main/java/com/hazelcast/internal/nearcache/impl/preloader/NearCachePreloader.java

```diff
@@ -0,0 +1,303 @@
+            logger.info(format("Loaded %d keys of Near Cache %s in %d ms", loadedKeys, nearCacheName, elapsedMillis));
+        } catch (Exception e) {
+            logger.warning(format("Could not pre-load Near Cache %s (%s): [%s] %s", nearCacheName, storeFile.getAbsolutePath(),
+                    e.getClass().getSimpleName(), e.getMessage()));
+        } finally {
+            tmpBytes = null;
+            closeResource(bis);
+        }
+    }
+
+    private boolean checkHeader(BufferingInputStream bis) throws IOException {
+        int magicBytes = readInt(bis);
+        if (magicBytes != MAGIC_BYTES) {
+            logger.warning(format("Found invalid header for Near Cache %s (%s)", nearCacheName, storeFile.getAbsolutePath()));
+            return false;
+        }
+        int fileFormat = readInt(bis);
+        if (fileFormat < 0 || fileFormat > FileFormat.values().length - 1) {
+            logger.warning(format("Found invalid file format for Near Cache %s (%s)", nearCacheName,
+                    storeFile.getAbsolutePath()));
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Stores the keys from the supplied {@link NearCacheRecordMap} instances.
+     *
+     * We need to support multiple records maps here, since Hazelcast Enterprise has a segmented
+     * {@link com.hazelcast.internal.nearcache.NearCacheRecordStore} which contains multiple records maps.
+     *
+     * @param records the {@link NearCacheRecordMap} instances to retrieve the keys from
+     * @param <R>     the {@link NearCacheRecord} type
+     * @param <NCRM>  the {@link NearCacheRecordMap} type
+     */
+    public <R extends NearCacheRecord, NCRM extends NearCacheRecordMap<KS, R>> void storeKeys(NCRM... records) {
+        long started = System.nanoTime();
+        FileOutputStream fos = null;
+        try {
+            tmpBytes = new byte[INT_SIZE_IN_BYTES];
+            lastWrittenBytes = 0;
+            lastKeyCount = 0;
+
+            fos = new FileOutputStream(tmpStoreFile, false);
+
+            // writer header
+            writeInt(fos, MAGIC_BYTES);
+            writeInt(fos, FileFormat.INTERLEAVED_LENGTH_FIELD.ordinal());
+
+            // writer keys
+            for (NCRM record : records) {
+                writeKeySet(fos, fos.getChannel(), record.keySet());
+            }
+
+            // cleanup if no keys have been written
+            if (lastKeyCount == 0) {
+                deleteQuietly(storeFile);
+                persistenceStats(started);
+                return;
+            }
+
+            fos.flush();
+            rename(tmpStoreFile, storeFile);
+
+            persistenceStats(started);
+        } catch (Exception e) {
+            logger.warning(format("Could not store keys of Near Cache %s (%s): [%s] %s", nearCacheName,
+                    storeFile.getAbsolutePath(), e.getClass().getSimpleName(), e.getMessage()));
+        } finally {
+            tmpBytes = null;
```

Comment:
I changed that code part several times. But you're right, I'll make it a final field. I think we can afford a few bytes being held forever.